rust/connlib/tunnel/src/device_channel/tun_windows.rs

```diff
@@ -24,48 +24,52 @@ use windows::Win32::{
     },
     Networking::WinSock::{AF_INET, AF_INET6},
 };
+use wintun::Adapter;
 
-// TODO: Double-check that all these get dropped gracefully on disconnect
-pub struct Tun {
+// Not sure how this and `TUNNEL_NAME` differ
+const ADAPTER_NAME: &str = "Firezone";
+
+pub(crate) struct Tun {
     /// The index of our network adapter, we can use this when asking Windows to add / remove routes / DNS rules
     /// It's stable across app restarts and I'm assuming across system reboots too.
     iface_idx: u32,
     packet_rx: mpsc::Receiver<wintun::Packet>,
-    _recv_thread: std::thread::JoinHandle<()>,
+    recv_thread: Option<std::thread::JoinHandle<()>>,
     session: Arc<wintun::Session>,
     routes: HashSet<IpNetwork>,
 }
 
 impl Drop for Tun {
     fn drop(&mut self) {
-        if let Err(e) = self.session.shutdown() {
-            tracing::error!("wintun::Session::shutdown: {e:#?}");
+        if let Err(error) = self.session.shutdown() {
+            tracing::error!(?error, "wintun::Session::shutdown");
+        }
+        if let Some(recv_thread) = self.recv_thread.take() {
+            // We must join the worker thread here to prevent issue #4765
+            if let Err(error) = recv_thread.join() {
+                tracing::error!(?error, "Couldn't join `recv_thread`");
+            }
+        } else {
+            tracing::error!("No `recv_thread` in `Tun`");
```

Comment:
we can just `self.recv_thread.take().unwrap().join();`  This should never fail and if it fails there's nothing we can do

Reply:
I want to at least log errors. I didn't read the docs close enough the first time, if `join` returns `Err` it's not that it didn't join, but that the worker thread panicked, and it returns the panic message. So I'll fix that and simplify the `if-let`